### ISSUE TITLE
Flatten is not constant folded properly by constfold grappler.

### ISSUE THREAD

**Issue by trevor-m** [tensorflow/tensorflow#25402]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): master (1/31/19)
- Python version: 3.5.2
- Bazel version (if compiling from source): 0.19.2
- GCC/Compiler version (if compiling from source): c++ (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
- CUDA/cuDNN version: CUDA 10.0.130
- GPU model and memory: Quadro RTX 6000, 24GiB

**Describe the current behavior**
tf.layers.Flatten uses Shape, StridedSlice, Pack ops to build the new shape for Reshape. These ops are not constant folded.

**Describe the expected behavior**
 Since the shape of `x` is known at compile time, the constant folder should be able to fold away these ops to create a single Const for the new shape for Reshape.

**Code to reproduce the issue**
```python
import tensorflow as tf
import math
from tensorflow.core.protobuf import meta_graph_pb2
from tensorflow.core.protobuf import rewriter_config_pb2
from tensorflow.python.grappler import tf_optimizer

def build_graph(x, output_name='output'):
  # x.shape = (None, 1, 1, 1536)
  x = tf.layers.Flatten()(x)
  # This should perform exactly the same as:
  # x = tf.reshape(x, [-1, 1536])
  return tf.identity(x, name=output_name)

def apply_constfold(frozen_graph, output_nodes):
  graph = tf.Graph()
  with graph.as_default():
    tf.import_graph_def(frozen_graph, name="")
  grappler_meta_graph_def = tf.train.export_meta_graph(graph_def=graph.as_graph_def(add_shapes=True), graph=graph)

  _to_bytes = lambda s: s.encode("utf-8", errors="surrogateescape")
  output_collection = meta_graph_pb2.CollectionDef()
  output_list = output_collection.node_list.value
  for i in output_nodes:
    if isinstance(i, tf.Tensor):
      output_list.append(_to_bytes(i.name))
    else:
      output_list.append(_to_bytes(i))
  # TODO(laigd): use another key as the outputs are really not train_op.
  grappler_meta_graph_def.collection_def["train_op"].CopyFrom(output_collection)
  rewriter_config = rewriter_config_pb2.RewriterConfig()
  rewriter_config.optimizers.extend(["constfold"])

  session_config_with_trt = tf.ConfigProto()
  session_config_with_trt.graph_options.rewrite_options.CopyFrom(
      rewriter_config)
  frozen_graph = tf_optimizer.OptimizeGraph(session_config_with_trt, grappler_meta_graph_def, graph_id=b"tf_graph")
  return frozen_graph

if __name__ == '__main__':
  with tf.Graph().as_default():
    # Create graph
    x = tf.placeholder(dtype=tf.float32, shape=(None, 1, 1, 1536), name='input')
    y = build_graph(x)
    # Initialize
    with tf.Session() as sess:
      sess.run(tf.global_variables_initializer())
      # Freeze graph
      frozen_graph = tf.graph_util.convert_variables_to_constants(
          sess,
          sess.graph_def,
          output_node_names=['output'])

  print('Nodes before:')
  [print(n.name, n.op) for n in frozen_graph.node]

  # const folding
  frozen_graph = apply_constfold(frozen_graph, output_nodes=['output'])

  print('----------------------------------------')
  print('Nodes after:')
  [print(n.name, n.op) for n in frozen_graph.node]
```

**Output**
```
Nodes before:
input Placeholder
flatten/Shape Shape
flatten/strided_slice/stack Const
flatten/strided_slice/stack_1 Const
flatten/strided_slice/stack_2 Const
flatten/strided_slice StridedSlice
flatten/Reshape/shape/1 Const
flatten/Reshape/shape Pack
flatten/Reshape Reshape
output Identity
-------------------------
Nodes after:
input Placeholder
flatten/strided_slice/stack Const
flatten/strided_slice/stack_1 Const
flatten/strided_slice/stack_2 Const
flatten/Reshape/shape/1 Const
flatten/Shape Shape
flatten/strided_slice StridedSlice
flatten/Reshape/shape Pack
flatten/Reshape Reshape
output Identity
```

** Expected Output **
```
Nodes before:
input Placeholder
flatten/Shape Shape
flatten/strided_slice/stack Const
flatten/strided_slice/stack_1 Const
flatten/strided_slice/stack_2 Const
flatten/strided_slice StridedSlice
flatten/Reshape/shape/1 Const
flatten/Reshape/shape Pack
flatten/Reshape Reshape
output Identity
-------------------------
Nodes after:
input Placeholder
Reshape/shape Const
Reshape Reshape
output Identity
```

**Comment by trevor-m** [comment]
@rmlarsen

**Comment by trevor-m** [comment]
@aaroey Could you take a look at this? This is affecting InceptionV4 in TF-TRT since the Flatten() layer at the end of the network is not able to be converted.

**Comment by aaroey** [comment]
@trevor-m sorry for being late. I can reproduce the problem, but I think it'll be hard to solve this problem in Grappler unless with explicit pattern matching. So, the output `flatten/strided_slice:0` is the batch dimension of the input (lets name it `batch_dim`), then the output of `flatten/Reshape/shape:0` is `[batch_dim, -1]` (note, not `[-1, 1536]` even though the outcome are same with this particular graph). Since `batch_dim` is not known at compile time, it won't be able to fold it, and I don't know any existing framework that can track the `batch_dim` and flow it along the edges.

@rmlarsen any suggestions?

**Comment by smit-hinsu** [comment]
+@andyly as well.

**Comment by andyly** [comment]
As @aaroey mentioned, because the shape is not fully defined for the PlaceHolder, it cannot be constant folded down into the Shape, resulting in no further constant folding. One option is to write an explicit transformation looking for this pattern, but I am not sure how generalized this sort of pattern can be.

**Comment by trevor-m** [comment]
Thanks everyone for investigating!
So currently `Flatten` is doing the following (simplified):
```
x = tf.reshape(x, shape=(x.shape[0], -1))
```
https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/python/keras/layers/core.py#L551-L553

Could we instead change flatten to do the following? basically, if all dims other than the batch dim are fully defined we can use `[-1, 1536]` instead of `[batch_dim, -1]`
```
if -1 not in x.shape[1:]:
  x = tf.reshape(x, shape=(-1, prod(x.shape[1:]))
else:
  x = tf.reshape(x, shape=(x.shape[0], -1))
```
